lib/Service/ImageService.php

```diff
@@ -0,0 +1,631 @@
+		}
+		$imageFile = $this->getFileFromPath($path, $userId);
+		$saveDir = $this->getAttachmentDirectoryForFile($textFile, true);
+		return $this->copyImageFile($imageFile, $saveDir, $textFile);
+	}
+
+	/**
+	 * @param File $imageFile
+	 * @param Folder $saveDir
+	 * @param File $textFile
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 */
+	private function copyImageFile(File $imageFile, Folder $saveDir, File $textFile): array {
+		$mimeType = $imageFile->getMimeType();
+		if (in_array($mimeType, ImageController::IMAGE_MIME_TYPES, true)) {
+			$fileName = (string) time() . '-' . $imageFile->getName();
+			$targetPath = $saveDir->getPath() . '/' . $fileName;
+			$targetFile = $imageFile->copy($targetPath);
+			// get file type and name
+			return [
+				'name' => $fileName,
+				'id' => $targetFile->getId(),
+				'documentId' => $textFile->getId(),
+			];
+		}
+		return [
+			'error' => 'Unsupported file type',
+		];
+	}
+
+	/**
+	 * Download and save an image from a link in the attachment folder
+	 *
+	 * @param int $documentId
+	 * @param string $link
+	 * @param string $userId
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws NotPermittedException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Lock\LockedException
+	 * @throws \OC\User\NoUserException
+	 */
+	public function insertImageLink(int $documentId, string $link, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
+			throw new NotPermittedException('No write permissions');
+		}
+		$saveDir = $this->getAttachmentDirectoryForFile($textFile, true);
+		return $this->downloadLink($saveDir, $link, $textFile);
+	}
+
+	/**
+	 * Download and save an image from a link in the attachment folder in a public context
+	 *
+	 * @param int|null $documentId
+	 * @param string $link
+	 * @param string $shareToken
+	 * @return array|string[]
+	 * @throws NotFoundException
+	 * @throws NotPermittedException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Lock\LockedException
+	 * @throws \OC\User\NoUserException
+	 */
+	public function insertImageLinkPublic(?int $documentId, string $link, string $shareToken): array {
+		if (!$this->hasUpdatePermissions($shareToken)) {
+			throw new NotPermittedException('No write permissions');
+		}
+		$textFile = $this->getTextFilePublic($documentId, $shareToken);
+		$saveDir = $this->getAttachmentDirectoryForFile($textFile, true);
+		return $this->downloadLink($saveDir, $link, $textFile);
+	}
+
+	/**
+	 * Check if the shared access has write permissions
+	 *
+	 * @param string $shareToken
+	 * @return bool
+	 */
+	private function hasUpdatePermissions(string $shareToken): bool {
+		try {
+			$share = $this->shareManager->getShareByToken($shareToken);
+			return ($share->getShareType() === IShare::TYPE_LINK && $share->getPermissions() & Constants::PERMISSION_UPDATE);
```

Comment:
`IShare::TYPE_EMAIL` would also be a case where this should work as they behave the same like share links.

Reply:
`IShare::TYPE_ROOM` would also be possible if the room is shared to the public, then such shares would be used to access the file as guests

Reply:
:+1: Done in fad15fa0a42010eab2905dba2e9479d1a53ca025